### PR TITLE
Fix diayn and derive GoalGenerator from RLAlgorithm

### PR DESCRIPTION
**Pull request by emailweixu**
Some new changes are not tested against diayn. Added it to train_play_test.

The generator need to drive from RLAlgorithm so that we use rl algorithm to train a policy to generate the goal.

### CODE REVIEW

File: alf/algorithms/goal_generator.py [hunk 1]
```diff
@@ -30,28 +31,40 @@
...
+class RandomCategoricalGoalGenerator(RLAlgorithm):
     """Random Goal Generation Module
 
     This module generates a random categorical goal for the agent
     in the beginning of every episode.
     """
 
-    def __init__(self, num_of_goals, name="RandomCategoricalGoalGenerator"):
+    def __init__(self,
+                 observation_spec,
```
Comment (Haichao-Zhang) [inline alf/algorithms/goal_generator.py]: For the ```observation_spec```here, I understand this is due to the change of the base class to ```RLAlgorithm```, but still it feels a bit unnatural to take something unused as input.

File: alf/algorithms/goal_generator.py [hunk 2]
```diff
@@ -86,6 +96,7 @@ def train_step(self,
             calc_goal_update (bool): perform goal alculattion the goal if true;
```
Comment (Haichao-Zhang) [inline alf/algorithms/goal_generator.py]: comments for ```calc_goal_update``` can be removed

File: alf/algorithms/goal_generator.py (hunk 1 above)
Comment (hnyu) [reply alf/algorithms/goal_generator.py]: #335

Comment (hnyu) [reply alf/algorithms/goal_generator.py]: > For the `observation_spec`here, I understand this is due to the change of the base class to `RLAlgorithm`, but still it feels a bit unnatural to take something unused as input.

Check out this issue #335

Comment (emailweixu) [reply alf/algorithms/goal_generator.py]: You actually use it at line 66.

**Pull request merged**